source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -428,6 +435,129 @@ int StreamHandleWrapper::luaLogCritical(lua_State* state) {
+  if (hash == "md4") {
+    return EVP_md4();
+  } else if (hash == "md5") {
+    return EVP_md5();
+  } else if (hash == "sha1") {
+    return EVP_sha1();
+  } else if (hash == "sha224") {
+    return EVP_sha224();
+  } else if (hash == "sha256") {
+    return EVP_sha256();
+  } else if (hash == "sha384") {
+    return EVP_sha384();
+  } else if (hash == "sha512") {
+    return EVP_sha512();
+  } else if (hash == "md5_sha1") {
+    return EVP_md5_sha1();
+  } else {
+    return nullptr;
+  }
+}
+
+int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
+  // Step 1: get public key
+  auto ptr = lua_touserdata(state, 2);
+  auto key = reinterpret_cast<EVP_PKEY*>(ptr);
+
+  // Step 2: initialize EVP_MD_CTX
+  EVP_MD_CTX* ctx = EVP_MD_CTX_new();
+
+  // Step 3: initialize EVP_MD
+  absl::string_view hash_func = luaL_checkstring(state, 3);
+  const EVP_MD* md = getDigest(hash_func);
+
+  if (md == nullptr) {
+    lua_pushboolean(state, false);
+    std::string err_msg = absl::StrCat(hash_func, " is not supported.");
+    lua_pushlstring(state, err_msg.data(), err_msg.length());
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  // Step 4: initialize EVP_DigestVerify
+  int ok = EVP_DigestVerifyInit(ctx, nullptr, md, nullptr, key);
+  if (!ok) {
+    lua_pushboolean(state, false);
+    absl::string_view err_msg = "Failed to initialize digest verify.";
+    lua_pushlstring(state, err_msg.data(), err_msg.length());
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  // Step 5: verify signature
+  const char* signature = luaL_checkstring(state, 4);
+  int sigLen = luaL_checknumber(state, 5);
+  absl::string_view sig(signature, sigLen);
+
+  const char* clearText = luaL_checkstring(state, 6);
+  int textLen = luaL_checknumber(state, 7);
+  absl::string_view text(clearText, textLen);
+
+  ok = EVP_DigestVerify(ctx, reinterpret_cast<const uint8_t*>(sig.data()), sig.length(),
+                        reinterpret_cast<const uint8_t*>(text.data()), text.length());
+
+  // Step 6: check result
+  if (ok == 1) {
+    lua_pushboolean(state, true);
+    lua_pushnil(state);
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  lua_pushboolean(state, false);
+  std::string err_msg = absl::StrCat("Failed to verify digest. Error code: ", ok);
+  lua_pushlstring(state, err_msg.data(), err_msg.length());
+  EVP_MD_CTX_free(ctx);
+  return 2;
+}
+
+int StreamHandleWrapper::luaDecodeBase64(lua_State* state) {
+  // Get input string
+  absl::string_view str = luaL_checkstring(state, 2);
+  std::string output;
+
+  bool ok = absl::Base64Unescape(str, &output);
+
+  if (ok) {
+    lua_pushlstring(state, output.data(), output.length());
+  } else {
+    // If decode failed, push a nil into stack.
+    lua_pushnil(state);
+  }
+
+  return 1;
+}
+
+int StreamHandleWrapper::luaImportPublicKey(lua_State* state) {
+  // Get byte array and the length.
+  const char* str = luaL_checkstring(state, 2);
+  int n = luaL_checknumber(state, 3);
```

Comment:
I think either @lizan or @mattklein123 should review the FFI for Lua, I'd just want to make sure we haven't done anything dumb there.

Reply:
I think @lizan is already in the pr.